src/libpmemfile/preload.c

```diff
@@ -0,0 +1,2024 @@
+	    (void *)where.at.pmem_fda.pool->pool, where.path, r);
+
+	return check_errno(r);
+}
+
+static long
+hook_chdir(const char *path)
+{
+	struct resolved_path where;
+
+	long result;
+
+	log_write("%s(\"%s\")", __func__, path);
+
+	util_rwlock_wrlock(&pmem_cwd_lock);
+
+	resolve_path(cwd_desc(), path, &where, RESOLVE_LAST_SLINK);
+
+	if (where.error_code != 0) {
+		result = where.error_code;
+	} else if (is_fda_null(&where.at.pmem_fda)) {
+		result = syscall_no_intercept(SYS_chdir, where.path);
+		if (result == 0)
+			cwd_pool = NULL;
+	} else {
+		if (cwd_pool != where.at.pmem_fda.pool) {
+			cwd_pool = where.at.pmem_fda.pool;
+			syscall_no_intercept(SYS_chdir, cwd_pool->mount_point);
+		}
+		if (pmemfile_chdir(cwd_pool->pool, where.path) == 0)
+			result = 0;
+		else
+			result = -errno;
+		log_write("pmemfile_chdir(%p, \"%s\") = %ld",
+		    cwd_pool->pool, where.path, result);
+		check_errno(result);
+	}
+
+	util_rwlock_unlock(&pmem_cwd_lock);
+
+	return result;
+}
+
+static long
+hook_fchdir(long fd)
+{
+	if (fd == AT_FDCWD)
+		return 0;
+
+	long result;
+
+	log_write("%s(\"%ld\")", __func__, fd);
+
+	util_rwlock_wrlock(&pmem_cwd_lock);
+
+	if (fd_pool_has_allocated(fd)) {
+		struct fd_association *where = fd_table + fd;
+		if (pmemfile_fchdir(where->pool->pool, where->file) == 0) {
+			cwd_pool = where->pool;
+			result = 0;
+		} else {
+			result = -errno;
+		}
+		log_write("pmemfile_fchdir(%p, %p) = %ld",
+		    where->pool->pool, where->file, result);
+		check_errno(result);
+	} else {
+		result = syscall_no_intercept(SYS_fchdir, fd);
+		if (result == 0)
+			cwd_pool = NULL;
+	}
+
+	util_rwlock_unlock(&pmem_cwd_lock);
+
+	return result;
+}
+
+static long
+hook_getcwd(char *buf, size_t size)
+{
+	if (cwd_pool == NULL)
+		return syscall_no_intercept(SYS_getcwd, buf, size);
+
+	size_t mlen = strlen(cwd_pool->mount_point);
+	if (mlen >= size)
+		return -ERANGE;
+	strcpy(buf, cwd_pool->mount_point);
+	if (pmemfile_getcwd(cwd_pool->pool, buf + mlen, size - mlen) != NULL)
+		return 0;
+	else
+		return check_errno(-errno);
+}
+
+static long log_fd = -1;
+
+static void
+log_init(const char *path, const char *trunc)
+{
+	if (path != NULL) {
+		int flags = O_CREAT | O_RDWR | O_APPEND | O_TRUNC;
+		if (trunc && trunc[0] == '0')
+			flags &= ~O_TRUNC;
+
+		log_fd = syscall_no_intercept(SYS_open, path, flags, 0600);
+	}
+}
+
+static void
+log_write(const char *fmt, ...)
+{
+	if (log_fd < 0)
+		return;
+
+	char buf[0x1000];
+	int len;
+	va_list ap;
+
+	va_start(ap, fmt);
+	len = vsnprintf(buf, sizeof(buf) - 1, fmt, ap);
+	va_end(ap);
+
+
+	if (len < 1)
+		return;
+
+	buf[len++] = '\n';
+
+	syscall_no_intercept(SYS_write, log_fd, buf, len);
+}
+
+static void
+open_new_pool(struct pool_description *p)
+{
+	if (p->pool == NULL) {
+		PMEMfilepool *pfp = pmemfile_pool_open(p->poolfile_path);
+		if (pmemfile_stat(pfp, "/", &p->pmem_stat))
+			FATAL("pmemfile_stat(\"/\") failed!");
+		__atomic_store_n(&p->pool, pfp, __ATOMIC_SEQ_CST);
```

Comment:
This might have made sense a hundred commits ago, but I guess it is just buggy right now. I'm going remove/rewrite pool handling related things.